### PR TITLE
MINIFICPP-1395 - Use Identifier instead of its stringified form wherever possible

### PR DESCRIPTION
**Pull request by adamdebreceni**
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/core/CoreComponentState.h [hunk 1]
```diff
@@ -48,13 +49,13 @@ class CoreComponentStateManagerProvider {
...
-  virtual std::shared_ptr<CoreComponentStateManager> getCoreComponentStateManager(const std::string& uuid) = 0;
+  virtual std::shared_ptr<CoreComponentStateManager> getCoreComponentStateManager(const utils::Identifier& uuid) = 0;
 
   virtual std::shared_ptr<CoreComponentStateManager> getCoreComponentStateManager(const CoreComponent& component) {
-    return getCoreComponentStateManager(component.getUUIDStr());
+    return getCoreComponentStateManager(component.getUUID());
   }
 
-  virtual std::unordered_map<std::string, std::unordered_map<std::string, std::string>> getAllCoreComponentStates() = 0;
+  virtual std::map<utils::Identifier, std::unordered_map<std::string, std::string>> getAllCoreComponentStates() = 0;
```
Comment (arpadboda) [inline libminifi/include/core/CoreComponentState.h:58]: What's the reason behind this change?
If you do this, unordered map include can be removed imho.

File: libminifi/include/sitetosite/SiteToSite.h [hunk 2]
```diff
@@ -246,24 +246,24 @@ class Transaction {
...
-
-    // Generate the global UUID for the transaction
-    uuid_str_ = uuid_.to_string();
   }
   // Destructor
   virtual ~Transaction() = default;
   // getUUIDStr
-  std::string getUUIDStr() {
-    return uuid_str_;
+  utils::SmallString<36> getUUIDStr() {
```
Comment (arpadboda) [inline libminifi/include/sitetosite/SiteToSite.h:253]: const

File: libminifi/src/CronDrivenSchedulingAgent.cpp [hunk 3]
```diff
@@ -56,8 +56,8 @@ utils::TaskRescheduleInfo CronDrivenSchedulingAgent::run(const std::shared_ptr<c
       } else {
         Bosma::Cron schedule(processor->getCronPeriod());
         result = schedule.cron_to_next(from);
-        last_exec_[uuidStr] = result;
-        schedules_.insert(std::make_pair(uuidStr, schedule));
+        last_exec_[uuid] = result;
+        schedules_.insert(std::make_pair(uuid, schedule));
```
Comment (arpadboda) [inline libminifi/src/CronDrivenSchedulingAgent.cpp:60]: It's old code, but it's a map! :)

schedules_[uuid] = schedule;

**Review by arpadboda — COMMENTED** [review]
Mostly looks good, some minor comments.

File: libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp [hunk 4]
```diff
@@ -29,20 +29,29 @@ PersistableKeyValueStoreService::PersistableKeyValueStoreService(const std::stri
...
-bool PersistableKeyValueStoreService::getImpl(std::unordered_map<std::string, std::string>& kvs) {
-  return get(kvs);
+bool PersistableKeyValueStoreService::getImpl(std::map<utils::Identifier, std::string>& kvs) {
+  std::unordered_map<std::string, std::string> states;
+  if (!get(states)) {
+    return false;
+  }
+  kvs.clear();
+  for (const auto& state : states) {
+    const auto uuid = utils::Identifier::parse(state.first);
```
Comment (fgerlits) [inline libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp:47]: it is not immediately clear to me from this code what kind of thing `uuid` is; I would either change `auto` to `utils::optional<Identifier>` or change the name of the variable to something like `optional_uuid`

File: libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp [hunk 5]
```diff
@@ -29,20 +29,29 @@ PersistableKeyValueStoreService::PersistableKeyValueStoreService(const std::stri
...
+    kvs[uuid.value()] = state.second;
```
Comment (fgerlits) [inline libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp:48]: what happens if this throws?  where do we catch it?

File: libminifi/src/core/reporting/SiteToSiteProvenanceReportingTask.cpp [hunk 6]
```diff
@@ -68,16 +68,29 @@ void setJsonStr(const std::string& key, const std::string& value, rapidjson::Val
...
+rapidjson::Value getStringValue(const std::string& value, rapidjson::Document::AllocatorType& alloc) {
   rapidjson::Value Val;
   Val.SetString(value.c_str(), value.length(), alloc);
   return Val;
 }
 
-void appendJsonStr(const std::string& value, rapidjson::Value& parent, rapidjson::Document::AllocatorType& alloc) { // NOLINT
+template<size_t N>
+rapidjson::Value getStringValue(const utils::SmallString<N>& value, rapidjson::Document::AllocatorType& alloc) {
+  rapidjson::Value Val;
```
Comment (fgerlits) [inline libminifi/src/core/reporting/SiteToSiteProvenanceReportingTask.cpp:79]: very minor, but I would rename this variable (and the other `valueVal`s) to `target_value` or `json_value` or something similar to match our naming conventions

File: libminifi/include/core/CoreComponentState.h (hunk 1 above)
Comment (adamdebreceni) [reply libminifi/include/core/CoreComponentState.h:58]: since it manages CoreComponents (it's in the name) and each CoreComponent has an id, it makes sense to use the id directly instead of `std::string`, if you are asking about the `unordered_map` => `map` change, I tried to avoid writing a hash implementation for the `Identifier` (which we then would have to manually specify each time we write the unordered_map type)

File: libminifi/src/provenance/Provenance.cpp [hunk 7]
```diff
@@ -346,25 +352,33 @@ bool ProvenanceEventRecord::DeSerialize(const uint8_t *buffer, const size_t buff
...
+      ret = outStream.read(childUUIDStr);
       if (ret <= 0) {
         return false;
       }
-      this->addChildUuid(childUUID);
+      utils::optional<utils::Identifier> childUUID = utils::Identifier::parse(childUUIDStr);
+      if (!childUUID) {
+        return false;
+      }
+      this->addChildUuid(childUUID.value());
```
Comment (fgerlits) [inline libminifi/src/provenance/Provenance.cpp:381]: we do this a lot; it could be pulled out to a function, or even maybe to a new `outStream.read(utils::Identifier&)` overload

File: libminifi/src/CronDrivenSchedulingAgent.cpp (hunk 3 above)
Comment (adamdebreceni) [reply libminifi/src/CronDrivenSchedulingAgent.cpp:60]: I think they have different semantics, your proposal overwrites if it already exists, the current version does not change the map if there was a same key in it, unclear if that was intentional (probably not) but I wanted to change semantics as little as possible

File: libminifi/include/core/CoreComponentState.h (hunk 1 above)
Comment (arpadboda) [reply libminifi/include/core/CoreComponentState.h:57]: Okay, makes sense, thanks!

File: libminifi/src/CronDrivenSchedulingAgent.cpp (hunk 3 above)
Comment (arpadboda) [reply libminifi/src/CronDrivenSchedulingAgent.cpp:60]: 👍

**Review by arpadboda — APPROVED** [review]

File: libminifi/test/TestBase.cpp [hunk 8]
```diff
@@ -104,9 +104,9 @@ std::shared_ptr<core::Processor> TestPlan::addProcessor(const std::shared_ptr<co
     connection->setDestination(processor);
 
     utils::Identifier uuid_copy, uuid_copy_next;
-    last->getUUID(uuid_copy);
+    uuid_copy = last->getUUID();
     connection->setSourceUUID(uuid_copy);
-    processor->getUUID(uuid_copy_next);
+    uuid_copy_next = processor->getUUID();
     connection->setDestinationUUID(uuid_copy_next);
```
Comment (fgerlits) [inline libminifi/test/TestBase.cpp:110]: minor, but this could be simplified to
```c++
connection->setSourceUUID(last->getUUID());
connection->setDestinationUUID(processor->getUUID());
```
(also further down in this file)

File: nanofi/src/cxx/Plan.cpp [hunk 9]
```diff
@@ -304,11 +304,11 @@ std::shared_ptr<minifi::Connection> ExecutionPlan::connectProcessors(std::shared
   connection->setSource(src_proc);
 
   utils::Identifier uuid_copy, uuid_copy_next;
-  src_proc->getUUID(uuid_copy);
+  uuid_copy = src_proc->getUUID();
   connection->setSourceUUID(uuid_copy);
```
Comment (fgerlits) [inline nanofi/src/cxx/Plan.cpp:308]: here, too, this could be `connection->setSourceUUID(src_proc->getUUID());` etc

File: libminifi/include/utils/SmallString.h [hunk 10]
```diff
@@ -34,6 +34,10 @@ class SmallString : public std::array<char, N + 1> {
     return {c_str()};
   }
 
+  size_t length() const noexcept {
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/utils/SmallString.h:37]: I intended to mimic the signature of `std::string::length` on purpose, to make it as similar as possible, and that is only `constexpr` from `c++20`, should I change it to `constexpr`?

File: libminifi/include/utils/ThreadPool.h [hunk 11]
```diff
@@ -45,6 +45,8 @@ namespace nifi {
 namespace minifi {
 namespace utils {
 
+using TaskId = std::string;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/utils/ThreadPool.h:48]: `GetTCP` creates a worker with `"workers"` name

https://github.com/apache/nifi-minifi-cpp/blob/d94e14f5fadb59a7b07da968b609d4f0831f4c27/extensions/standard-processors/processors/GetTCP.cpp#L290

File: libminifi/include/utils/SmallString.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (fgerlits) [reply libminifi/include/utils/SmallString.h:37]: I would, as it does not affect its usage, but I don't insist

File: libminifi/include/utils/ThreadPool.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (fgerlits) [reply libminifi/include/utils/ThreadPool.h:48]: I see, OK

**Review by szaszm — COMMENTED** [review]
Awesome changes, some questions and suggestions below.

File: libminifi/include/core/logging/Logger.h [hunk 12]
```diff
@@ -59,11 +59,18 @@ inline char const* conditional_conversion(const utils::SmallString<N>& arr) {
...
+    std::is_arithmetic<T>::value ||
+    std::is_enum<T>::value ||
+    std::is_pointer<T>::value>::type>
 inline T conditional_conversion(T const& t) {
   return t;
 }
 
+inline char const* conditional_conversion(const char* str) {
+  return str;
+}
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/logging/Logger.h:72]: for some reason I cannot reply to your comment about `std::hash` specialization:
from [google guidelines](https://google.github.io/styleguide/cppguide.html#std_hash)

> Do not define specializations of std::hash.

File: libminifi/src/ThreadedSchedulingAgent.cpp [hunk 13]
```diff
@@ -110,16 +110,15 @@ void ThreadedSchedulingAgent::schedule(std::shared_ptr<core::Processor> processo
...
 
 void ThreadedSchedulingAgent::stop() {
   SchedulingAgent::stop();
   std::lock_guard<std::mutex> lock(mutex_);
-  for (const auto& p : processors_running_) {
-    logger_->log_error("SchedulingAgent is stopped before processor was unscheduled: %s", p);
-    thread_pool_.stopTasks(p);
+  for (const auto& processor_id : processors_running_) {
+    logger_->log_error("SchedulingAgent is stopped before processor was unscheduled: %s", processor_id.to_string());
+    thread_pool_.stopTasks(processor_id.to_string());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/ThreadedSchedulingAgent.cpp:121]: yes that seems to be the case, (so processors with greater than one maxConcurrentTasks share the same `task_status_` flag)

File: libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp [hunk 14]
```diff
@@ -29,20 +29,31 @@ PersistableKeyValueStoreService::PersistableKeyValueStoreService(const std::stri
...
+  std::unordered_map<std::string, std::string> states;
+  if (!get(states)) {
+    return false;
+  }
+  kvs.clear();
+  for (const auto& state : states) {
+    utils::optional<utils::Identifier> optional_uuid = utils::Identifier::parse(state.first);
+    if (optional_uuid) {
+      kvs[optional_uuid.value()] = state.second;
+    }
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp:50]: since a `PerisistableKeyValueStoreService (PKVSS)` ISA `KeyValueStoreService (KVSS)` (publicly inherits) which can store arbitrary `std::string` keys, it is not impossible to cast a `PKVSS` to a `KVSS`, add some "invalid" key, and "mess up" the `PKVSS`, I would refactor this to not inherit from `KVSS` but contain an instance of it, for now I will add a log message

File: libminifi/include/core/logging/Logger.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/core/logging/Logger.h:72]: for this comment, it is needed so we can allow [this](https://godbolt.org/z/9drc8f) to compile, and now that I wrote this down, I realize I could simply remove the `const&` qualifier on the template

```
#include <iostream>

struct A{
  static constexpr const char* STR = "abcd";
};

template<typename T, typename = typename std::enable_if<
    std::is_arithmetic<T>::value ||
    std::is_enum<T>::value ||
    std::is_pointer<T>::value>::type>
inline T conditional_conversion(T const& t) {
  return t;
}

// inline char const* conditional_conversion(const char* str) {
//   return str;
// }

int main(){
  std::cout << conditional_conversion(A::STR) << std::endl;
}
```

File: libminifi/include/controllers/keyvalue/AbstractCoreComponentStateManagerProvider.h [hunk 15]
```diff
@@ -59,10 +60,10 @@ class AbstractCoreComponentStateManagerProvider : public std::enable_shared_from
...
 
  protected:
-  virtual bool setImpl(const std::string& key, const std::string& value) = 0;
-  virtual bool getImpl(const std::string& key, std::string& value) = 0;
-  virtual bool getImpl(std::unordered_map<std::string, std::string>& kvs) = 0;
-  virtual bool removeImpl(const std::string& key) = 0;
+  virtual bool setImpl(const utils::Identifier& key, const std::string& value) = 0;
+  virtual bool getImpl(const utils::Identifier& key, std::string& value) = 0;
+  virtual bool getImpl(std::map<utils::Identifier, std::string>& kvs) = 0;
+  virtual bool removeImpl(const utils::Identifier& key) = 0;
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/controllers/keyvalue/AbstractCoreComponentStateManagerProvider.h]: renamed the parameter, also added type alias for the `CoreComponentState` to `unordered_map<string, string>`

File: libminifi/src/ThreadedSchedulingAgent.cpp [hunk 16]
```diff
@@ -110,16 +110,15 @@ void ThreadedSchedulingAgent::schedule(std::shared_ptr<core::Processor> processo
...
+  processors_running_.insert(processor->getUUID());
 }
 
 void ThreadedSchedulingAgent::stop() {
   SchedulingAgent::stop();
   std::lock_guard<std::mutex> lock(mutex_);
   for (const auto& p : processors_running_) {
-    logger_->log_error("SchedulingAgent is stopped before processor was unscheduled: %s", p);
-    thread_pool_.stopTasks(p);
+    logger_->log_error("SchedulingAgent is stopped before processor was unscheduled: %s", p.to_string());
```
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/src/ThreadedSchedulingAgent.cpp]: done

File: libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp (hunk 4 above)
Comment (adamdebreceni) [reply libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp]: done

File: libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp (hunk 5 above)
Comment (adamdebreceni) [reply libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp]: rewrote it, to ignore such cases, but log errors, (should be refactored to contain an instance of `KeyValueStoreService` instead of extending it)

File: libminifi/src/provenance/Provenance.cpp [hunk 17]
```diff
@@ -128,7 +128,7 @@ bool ProvenanceEventRecord::Serialize(org::apache::nifi::minifi::io::BufferStrea
     return false;
   }
 
-  ret = outStream.write(this->flow_uuid_);
+  ret = outStream.write(this->flow_uuid_.to_string());
```
Comment (fgerlits) [inline libminifi/src/provenance/Provenance.cpp:131]: these `.to_string()`s are no longer needed here, right?

File: nanofi/src/cxx/Plan.cpp (hunk 9 above)
Comment (adamdebreceni) [reply nanofi/src/cxx/Plan.cpp]: done

File: libminifi/test/TestBase.cpp (hunk 8 above)
Comment (adamdebreceni) [reply libminifi/test/TestBase.cpp]: done

File: libminifi/src/provenance/Provenance.cpp (hunk 7 above)
Comment (adamdebreceni) [reply libminifi/src/provenance/Provenance.cpp]: added, also added an `OutputStream:read` overload for symmetry

**Review by fgerlits — APPROVED** [review]

File: libminifi/include/core/logging/Logger.h (hunk 12 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/core/logging/Logger.h]: The reason you couldn't reply to my comment about hash is because it was a reply to this discussion: https://github.com/apache/nifi-minifi-cpp/pull/932#discussion_r513307831

To this thread: acknowledged, I'm fine with either way.

File: libminifi/src/ThreadedSchedulingAgent.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/ThreadedSchedulingAgent.cpp:121]: @arpadboda could you confirm or clarify, as the author/major refactorer of this code?

File: libminifi/include/core/CoreComponentState.h (hunk 1 above)
Comment (szaszm) [reply libminifi/include/core/CoreComponentState.h]: marking as resolved again as per https://github.com/apache/nifi-minifi-cpp/pull/932#discussion_r514057657

File: libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/src/controllers/keyvalue/PersistableKeyValueStoreService.cpp:54]: Sounds like a violation of the Liskov substitution principle. Thanks for the mitigation.

**Review by szaszm — APPROVED** [review]

File: libminifi/include/utils/SmallString.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (adamdebreceni) [reply libminifi/include/utils/SmallString.h]: made it constexpr

File: libminifi/src/ThreadedSchedulingAgent.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (arpadboda) [reply libminifi/src/ThreadedSchedulingAgent.cpp:121]: I think it's intentional. 
The IDs are only used to identify the tasks belonging to a given component, the threadpool otherwise doesn't care. 
I wouldn't call it a beautiful design, but works as expected, so I kept it when the code was refactored.
